aten/src/ATen/native/cpu/UpSampleKernelAVXAntialias.h

```diff
@@ -422,97 +526,139 @@ void ImagingResampleHorizontalConvolution8u4x(
-          _mm_loadl_epi64((__m128i*)&lineIn0_min[i])),
-          _mm_loadl_epi64((__m128i*)&lineIn1_min[i]), 1);
-
+          _mm_loadl_epi64((__m128i *) (lineIn0_min + stride * i))),
+          _mm_loadl_epi64((__m128i *) (lineIn1_min + stride * i)), 1);
       // Apply mask_low:
-      //   [r0 g0 b0 a0  r1 g1 b1 a1  ... | R0 G0 B0 A0  R1 G1 B1 A1 ... ] ->
+      // RGBA:
       //   [r0 0 r1 0  g0 0 g1 0  b0 0 b1 0  a0 0 a1 0 | R0 0 R1 0  G0 0 G1 0  B0 0 B1 0  A0 0 A1 0]
+      // RGB:
+      //   [r0 0 r1 0  g0 0 g1 0  b0 0 b1 0  0 0 0 0 | R0 0 R1 0  G0 0 G1 0  B0 0 B1 0  0 0 0 0]
       auto pix1 = _mm256_shuffle_epi8(source1, mask_low);
       // Compute output value as C += w0 * C0 + w1 * C1 for each channel in 32-bit precision
       sss0 = _mm256_add_epi32(sss0, _mm256_madd_epi16(pix1, mmk));
 
       // Same as above for lines 2 and 3:
       auto source2 = _mm256_inserti128_si256(_mm256_castsi128_si256(
-          _mm_loadl_epi64((__m128i*)&lineIn2_min[i])),
-          _mm_loadl_epi64((__m128i*)&lineIn3_min[i]), 1);
+          _mm_loadl_epi64((__m128i *) (lineIn2_min + stride * i))),
+          _mm_loadl_epi64((__m128i *) (lineIn3_min + stride * i)), 1);
       auto pix2 = _mm256_shuffle_epi8(source2, mask_low);
       sss1 = _mm256_add_epi32(sss1, _mm256_madd_epi16(pix2, mmk));
     }
 
     // block 1
-    for (; i < ids_size; i++) {
+    const auto i32_aligned = num_channels == 4;
+    for (; i < ids_size - 1; i++) {
       // Load 1 value from weight vector
       // mmk = [wl_0 wh_0 0 0  wl_0 wh_0 0 0  ...]
       const auto mmk = _mm256_set1_epi32(k[i]);
 
       // Load 2 pixels (one per line) from input lines 0 and 1:
-      // source = [
-      //   r0 g0 b0 a0  0 0 0 0  0 0 0 0  0 0 0 0
-      //   R0 G0 B0 A0  0 0 0 0  0 0 0 0  0 0 0 0
+      // RGBA: pix1 = [
+      //   r0 0 0 0  g0 0 0 0  b0 0 0 0  a0 0 0 0
+      //   R0 0 0 0  G0 0 0 0  B0 0 0 0  A0 0 0 0
+      // ]
+      // RGB: pix1 = [
+      //   r0 0 0 0  g0 0 0 0  b0 0 0 0  r1 0 0 0
+      //   R0 0 0 0  G0 0 0 0  B0 0 0 0  R1 0 0 0
       // ]
       auto pix1 = _mm256_inserti128_si256(_mm256_castsi128_si256(
-          mm_cvtepu8_epi32(&lineIn0_min[i])),
-          mm_cvtepu8_epi32(&lineIn1_min[i]), 1);
+          mm_cvtepu8_epi32(lineIn0_min + stride * i, i32_aligned)),
+          mm_cvtepu8_epi32(lineIn1_min + stride * i, i32_aligned), 1);
       // Compute output value as C += w0 * C0 for each channel in 32-bit precision
       sss0 = _mm256_add_epi32(sss0, _mm256_madd_epi16(pix1, mmk));
 
       // Same as above for lines 2 and 3
       auto pix2 = _mm256_inserti128_si256(_mm256_castsi128_si256(
-          mm_cvtepu8_epi32(&lineIn2_min[i])),
-          mm_cvtepu8_epi32(&lineIn3_min[i]), 1);
+          mm_cvtepu8_epi32(lineIn2_min + stride * i, i32_aligned)),
+          mm_cvtepu8_epi32(lineIn3_min + stride * i, i32_aligned), 1);
+      sss1 = _mm256_add_epi32(sss1, _mm256_madd_epi16(pix2, mmk));
+    }
+
+    if (i == ids_size - 1) {
```

Comment:
> Is this condition check needed? I.e. isn't i always equal to ids_size - 1 here?

It can happen for C=4 that the block 2 can finish all the work and so we do not need to run this last index value

Reply:
> Also can you help me understand why we need to add some extra logic for the last element here? I would have assumed that if we needed some extra logic, it would only be needed for the C==3 case?

yes, this extra step is required by C=3 case and it was generalized to C=4 and C=3 cases.